mayastor/src/target/iscsi.rs

```diff
@@ -82,101 +88,70 @@ thread_local! {
+    if let Err(e) = create_portal_group(address, ISCSI_PORT_BE, ISCSI_PORTAL_GROUP_BE) {
+        return Err(e);
     }
-    unsafe {
-        let p = spdk_iscsi_portal_create(
-            portal_host.as_ptr(),
-            portal_port.as_ptr(),
-        );
-        if p.is_null() {
-            spdk_iscsi_portal_grp_release(pg);
-            return Err(Error::CreatePortal {});
-        }
-        spdk_iscsi_portal_grp_add_portal(pg, p);
-        if spdk_iscsi_portal_grp_open(pg) != 0 {
-            spdk_iscsi_portal_grp_release(pg);
-            return Err(Error::AddPortal {});
-        }
-        if spdk_iscsi_portal_grp_register(pg) != 0 {
-            spdk_iscsi_portal_grp_release(pg);
-            return Err(Error::RegisterPortalGroup {});
-        }
+    if let Err(e) = create_portal_group(address, ISCSI_PORT_FE, ISCSI_PORTAL_GROUP_FE) {
+        destroy_portal_group(ISCSI_PORTAL_GROUP_BE);
+        return Err(e);
     }
-    debug!("Created default iscsi portal group");
-
-    unsafe {
-        if spdk_iscsi_init_grp_create_from_initiator_list(
-            0,
-            1,
-            &mut (initiator_host.as_ptr() as *mut c_char) as *mut _,
-            1,
-            &mut (initiator_netmask.as_ptr() as *mut c_char) as *mut _,
-        ) != 0
-        {
-            spdk_iscsi_portal_grp_release(pg);
-            return Err(Error::CreateInitiatorGroup {});
-        }
+    if let Err(e) = create_initiator_group(ISCSI_INITIATOR_GROUP) {
+        destroy_portal_group(ISCSI_PORTAL_GROUP_BE);
+        destroy_portal_group(ISCSI_PORTAL_GROUP_FE);
+        return Err(e);
     }
     ADDRESS.with(move |addr| {
         *addr.borrow_mut() = Some(address.to_owned());
     });
-    debug!("Created default iscsi initiator group");
+    debug!("Created default iscsi initiator group and portal groups for address {}", address);
 
     Ok(())
 }
 
-/// Destroy iscsi default portal and initiator group.
+/// Destroy iscsi portal and initiator groups.
+fn destroy_iscsi_groups() {
+    destroy_initiator_group(ISCSI_INITIATOR_GROUP);
+    destroy_portal_group(ISCSI_PORTAL_GROUP_FE);
+    destroy_portal_group(ISCSI_PORTAL_GROUP_BE);
+}
+
 pub fn fini() {
-    unsafe {
-        let ig = spdk_iscsi_init_grp_unregister(0);
-        if !ig.is_null() {
-            spdk_iscsi_init_grp_destroy(ig);
-        }
-        let pg = spdk_iscsi_portal_grp_unregister(0);
-        if !pg.is_null() {
-            spdk_iscsi_portal_grp_release(pg);
-        }
-    }
+    destroy_iscsi_groups();
 }
 
-/// Export given bdev over iscsi. That involves creating iscsi target and
-/// adding the bdev as LUN to it.
-pub fn share(uuid: &str, bdev: &Bdev) -> Result<()> {
-    let iqn = target_name(uuid);
+fn share_as_iscsi_target(bdev_name: &str, bdev: &Bdev, pg_idx: c_int, ig_idx: c_int ) -> Result<String ,Error>{
+
+    let iqn = target_name(bdev_name);
     let c_iqn = CString::new(iqn.clone()).unwrap();
-    let mut group_idx: c_int = 0;
+    let mut portal_group_idx = pg_idx;
```

Comment:
Why copy these values?

Reply:
Agreed, I'll fix it.

Reply:
Fixed in commit https://github.com/openebs/Mayastor/commit/8fe37c12e6ad29c1c6c7d94da22f9d67a47a026b